Assert the assumption that two VM fields are loadable as a pair

ast-hugger opened the pull request:
#### 8268f5fdf8ebf9331602943fe32281afd03984ff
<pre>
Assert the assumption that two VM fields are loadable as a pair
<a href="https://bugs.webkit.org/show_bug.cgi?id=299933">https://bugs.webkit.org/show_bug.cgi?id=299933</a>
<a href="https://rdar.apple.com/161707462">rdar://161707462</a>

Reviewed by Mark Lam and Justin Michaud.

Fields `VM::topCallFrame` and `VM::topEntryFrame` are loaded and stored as a pair by the
assembly code. This PR adds a static assert to verify that the fields are laid out
accordingly, and a comment to point out that some loads and stores of `VM::topEntryFrame`
are implicit in pair instructions operating on `VM::topCallFrame`. A similar comment is
added to `VMEntryRecord`, which already has a static assert for its pair of fields used
with pair instructions.

The patch does not touch executable code and is not directly testable.

Canonical link: <a href="https://commits.webkit.org/300898@main">https://commits.webkit.org/300898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/521befeba0157f7595bdd2fe950c550ebc844418

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43733 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34430 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130845 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/76171 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dd05600f-955d-4e9f-be3a-daea30c8e995) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44457 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52327 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94342 "Failed to checkout and rebase branch from PR 51610") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62598 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6a796257-8f57-4f05-93d1-621e626a855d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110943 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74935 "Found 1 new API test failure: TestWTF:WTF_Condition.OneProducerOneConsumerOneSlot (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a54efd2e-2ae4-49c9-af33-da041636ecd6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34377 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29108 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74327 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116177 "Built successfully and passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105159 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29326 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133516 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122550 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50971 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38828 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102809 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51345 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107162 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102626 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running run-api-tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26154 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47977 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26229 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47846 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running compile-webkit") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50824 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56588 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153646 "Built successfully") | | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39051 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53644 "Built successfully") | | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/interpreter/VMEntryRecord.h:
```diff
@@ -41,6 +41,7 @@ struct VMEntryRecord {
      * after callee save registers where local variables would go.
      */
     VM* const m_vm;
+    // NOTE: the following two fields must stay together in this order because they are loaded/stored as a pair.
```
The better way to do this is to static_assert that `OFFSETOF(VMEntryRecord, m_prevTopCallFrame) + sizeof(void*) == OFFSETOF(VMEntryRecord, m_prevTopEntryFrame)`, and attach this comment / explanation to that static_assert.  That way, the compiler will tell us when someone breaks this invariant.

MenloDorian commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -378,6 +378,7 @@ class VM : public ThreadSafeRefCountedWithSuppressingSaferCPPChecking<VM>, publi
     // topEntryFrame.
     // FIXME: This should be a void*, because it might not point to a CallFrame.
     // https://bugs.webkit.org/show_bug.cgi?id=160441
+    // NOTE: the following two fields must stay together in this order because they are loaded/stored as a pair.
```
Ditto but using `OFFSETOF(VM, ...` instead.

MenloDorian reviewed (commented):
r=me with a suggestion for improvement: it would be good if the static assert comment points to the asm function / label where the pair access is used.  That saves the reader from having to search for it.  Alternatively, add the asm function / label to the commit message.

justinmichaud commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -378,6 +378,8 @@ class VM : public ThreadSafeRefCountedWithSuppressingSaferCPPChecking<VM>, publi
     // FIXME: This should be a void*, because it might not point to a CallFrame.
     // https://bugs.webkit.org/show_bug.cgi?id=160441
+    // The following two fields are sometimes treated as a pair in assembly code, making usages of the second one implicit.
+    // To find them, look for loadpairq/storepairq of "VM::topCallFrame" in *.asm files.
```
This might be a bit verbose for my taste, but I'm not sure. Maybe just the first line?

justinmichaud replied:
I see now that Mark already commented, we can stick with this

ast-hugger committed:
Assert the assumption that two VM fields are loadable as a pair
https://bugs.webkit.org/show_bug.cgi?id=299933
rdar://161707462

Reviewed by Mark Lam and Justin Michaud.

Fields `VM::topCallFrame` and `VM::topEntryFrame` are loaded and stored as a pair by the
assembly code. This PR adds a static assert to verify that the fields are laid out
accordingly, and a comment to point out that some loads and stores of `VM::topEntryFrame`
are implicit in pair instructions operating on `VM::topCallFrame`. A similar comment is
added to `VMEntryRecord`, which already has a static assert for its pair of fields used
with pair instructions.

The patch does not touch executable code and is not directly testable.

Canonical link: https://commits.webkit.org/300898@main
